lib/fluent/plugin/out_exec.rb

```diff
@@ -18,97 +18,82 @@
+
+    config_section :format do
+      config_set_default :@type, 'tsv'
     end
-    desc 'The name of the key to use as the event tag. This replaces the value in the event record.'
-    config_param :tag_key, :string, default: nil
-    desc 'The name of the key to use as the event time. This replaces the the value in the event record.'
-    config_param :time_key, :string, default: nil
-    desc 'The format for event time used when the time_key parameter is specified. The default is UNIX time (integer).'
-    config_param :time_format, :string, default: nil
-    desc "The format used to map the incoming events to the program input. (#{Fluent::ExecUtil::SUPPORTED_FORMAT.keys.join(',')})"
-    config_param :format, default: :tsv, skip_accessor: true do |val|
-      f = Fluent::ExecUtil::SUPPORTED_FORMAT[val]
-      raise Fluent::ConfigError, "Unsupported format '#{val}'" unless f
-      f
+
+    config_section :inject do
+      config_set_default :time_type, :string
+      config_set_default :localtime, false
     end
-    config_param :localtime, :bool, default: false
-    config_param :timezone, :string, default: nil
 
-    def compat_parameters_default_chunk_key
-      'time'
+    config_section :buffer do
+      config_set_default :delayed_commit_timeout, 300 # 5 min
     end
 
-    def configure(conf)
-      compat_parameters_convert(conf, :buffer, default_chunk_key: 'time')
+    attr_reader :formatter # for tests
 
+    def configure(conf)
+      compat_parameters_convert(conf, :inject, :formatter, :buffer, default_chunk_key: 'time')
       super
-
-      @formatter = case @format
-                   when :tsv
-                     if @keys.empty?
-                       raise Fluent::ConfigError, "keys option is required on exec output for tsv format"
-                     end
-                     Fluent::ExecUtil::TSVFormatter.new(@keys)
-                   when :json
-                     Fluent::ExecUtil::JSONFormatter.new
-                   when :msgpack
-                     Fluent::ExecUtil::MessagePackFormatter.new
-                   end
-
-      if @time_key
-        if @time_format
-          tf = Fluent::TimeFormatter.new(@time_format, @localtime, @timezone)
-          @time_format_proc = tf.method(:format)
-        else
-          @time_format_proc = Proc.new { |time| time.to_s }
-        end
-      end
+      @formatter = formatter_create
     end
 
     def format(tag, time, record)
-      out = ''
-      if @time_key
-        record[@time_key] = @time_format_proc.call(time)
-      end
-      if @tag_key
-        record[@tag_key] = tag
-      end
-      @formatter.call(record, out)
-      out
-    end
-
-    def write(chunk)
-      if chunk.respond_to?(:path)
-        prog = "#{@command} #{chunk.path}"
+      record = inject_values_to_record(tag, time, record)
+      if @formatter.formatter_type == :text_per_line
+        @formatter.format(tag, time, record).chomp + "\n"
```

Comment:
Using constant value for "\n"  is better for reducing temporary object allocation.